### PR TITLE
Fix migrating settings from plugin 0.20 to 0.21+

### PR DESCRIPTION
**Pull request by samrocketman**
fixes [JENKINS-29473][JENKINS-29473].  Authorization breaks when migrating from plugin 0.20 or below to 0.21 and above.  This is due to OAuth scopes not existing in prior configurations.  This patch fixes that.

[JENKINS-29473]: https://issues.jenkins-ci.org/browse/JENKINS-29473

### CODE REVIEW

**Comment by samrocketman** [comment]
@cloudbeesci code review, please :smile:.

**Comment by samrocketman** [comment]
cc @jhoblitt code review, please.

File: src/main/java/org/jenkinsci/plugins/GithubSecurityRealm.java [hunk 1]
```diff
@@ -663,6 +662,31 @@ public GroupDetails loadGroupByGroupname(String groupName)
...
+    /*
+       Migrate settings from 0.20 to 0.21+
+     */
+    @Extension
+    public static final class Migrator extends ItemListener {
+        @SuppressWarnings("deprecation")
+        @Override
+        public void onLoaded() {
+            try {
+            Jenkins instance = Jenkins.getInstance();
```
Comment (samrocketman) [inline src/main/java/org/jenkinsci/plugins/GithubSecurityRealm.java]: Need to fix indenting.

**Comment by samrocketman** [comment]
Fixed indenting and amended commit.

File: src/main/java/org/jenkinsci/plugins/GithubSecurityRealm.java [hunk 2]
```diff
@@ -72,17 +82,6 @@ of this software and associated documentation files (the "Software"), to deal
...
-import java.io.IOException;
-import java.lang.reflect.InvocationTargetException;
-import java.lang.reflect.Method;
-import java.net.InetSocketAddress;
-import java.net.Proxy;
-import java.util.HashSet;
-import java.util.Set;
-import java.util.logging.Logger;
-
-import static java.util.logging.Level.*;
```
Comment (jhoblitt) [inline src/main/java/org/jenkinsci/plugins/GithubSecurityRealm.java]: Looks like this was the only file importing `java.util.logging.Level.*`

File: src/main/java/org/jenkinsci/plugins/GithubSecurityRealm.java [hunk 3]
```diff
@@ -663,6 +662,31 @@ public GroupDetails loadGroupByGroupname(String groupName)
...
+    @Extension
+    public static final class Migrator extends ItemListener {
+        @SuppressWarnings("deprecation")
+        @Override
+        public void onLoaded() {
+            try {
+                Jenkins instance = Jenkins.getInstance();
+                if(instance.getSecurityRealm() instanceof GithubSecurityRealm) {
+                    GithubSecurityRealm myRealm = (GithubSecurityRealm) instance.getSecurityRealm();
+                    if(myRealm.getOauthScopes() == null || myRealm.getOauthScopes() == "") {
```
Comment (jhoblitt) [inline src/main/java/org/jenkinsci/plugins/GithubSecurityRealm.java]: I'm not sure about the empty string test.  I think the only way the `GithubSecurityRealm.oauthScopes` field would be `""` is if the user had already set it?

File: src/main/java/org/jenkinsci/plugins/GithubSecurityRealm.java [hunk 4]
```diff
@@ -663,6 +662,31 @@ public GroupDetails loadGroupByGroupname(String groupName)
...
+                        GithubSecurityRealm newRealm = new GithubSecurityRealm(myRealm.getGithubWebUri(), myRealm.getGithubApiUri(), myRealm.getClientID(), myRealm.getClientSecret());
```
Comment (jhoblitt) [inline src/main/java/org/jenkinsci/plugins/GithubSecurityRealm.java]: Is there a defacto line length limit used in the jenkins community?

Comment (samrocketman) [inline src/main/java/org/jenkinsci/plugins/GithubSecurityRealm.java]: Not sure.

File: src/main/java/org/jenkinsci/plugins/GithubSecurityRealm.java (hunk 3 above)
Comment (samrocketman) [inline src/main/java/org/jenkinsci/plugins/GithubSecurityRealm.java]: Makes sense.  I'll update it and test.

**Comment by samrocketman** [comment]
@jhoblitt I went ahead and removed checking for an empty string.  Once that verifies I'm going to do another plugin release as 0.21.2.  This problem will quickly exacerbate as more users attempt to upgrade so I want the patch to go out quick.  There's been one duplicate issue created about it in JIRA.  Thanks for your review :smile:.

If there's other code style changes or improvements we could always open a new pull request improving the style or length of lines.  I'm open to aesthetic and cleanup changes.

**Comment by samrocketman** [comment]
I have released plugin 0.21.2 to address this migration issue as quickly as possible.

**Pull request merged**